### PR TITLE
Added article about self contained building

### PR DESCRIPTION
**Pull request by schaefi**
Documentation

The article describes how to make use of the dice project to control and maintain image
build processes in a contained environment

### CODE REVIEW

**Comment by schaefi** [comment]
@tomschr if you have a minute it would great if you can review my attempts to write documentation :smile: 

I'm sure there is a lot to improve

Thanks much

File: doc/source/container_builder.rst [hunk 1]
```diff
@@ -0,0 +1,276 @@
+Building in a self-contained Environment
+========================================
+
+.. topic:: Abstract
+
+   This document describes how to build with KIWI using
+   dice as a container buildsystem in order to eliminate
+   any build host system requirements except the ones
+   dice pulls in
```
Comment (tomschr) [inline doc/source/container_builder.rst]: Missing dot. :-)

File: doc/source/container_builder.rst [hunk 2]
```diff
@@ -0,0 +1,276 @@
...
+
+Requirements
+------------
+
+The following components are needed in order to build in a
```
Comment (tomschr) [inline doc/source/container_builder.rst]: to build in a -> to build a (remove "in"?)

File: doc/source/container_builder.rst (hunk 1 above)
Comment (tomschr) [inline doc/source/container_builder.rst]: Maybe also mention the target group. Which users should read this article?

File: doc/source/container_builder.rst [hunk 3]
```diff
@@ -0,0 +1,276 @@
...
+  container support in the kernel.
+
+* docker build container for KIWI.
+
+* optionally :command:`vagrant` - a framework to run provision
+  and control virtual machines and container instances. vagrant
+  has a very nice way to provision a machine prior to running the
+  actual build. vagrant also supports docker as a provider which
+  makes it a perfect fit for complex provisioning tasks in
+  combination with the docker container system.
```
Comment (tomschr) [inline doc/source/container_builder.rst]: Some repetition. Maybe rephrase it like this:

```
* optionally :command:`vagrant` -  a framework to run provision* and control virtual
  machines and container instances. vagrant has a very nice way to provision a
  machine prior to running the actual build. It also supports docker as a provider which
  makes it a perfect fit for complex provisioning tasks in combination with the docker
  container system.
```

* I'm not sure what you mean by "run provision".

File: doc/source/container_builder.rst [hunk 4]
```diff
@@ -0,0 +1,276 @@
...
+
+dice packages are available at the respective buildservice project
+at http://download.opensuse.org/repositories/Virtualization:/Appliances. 
+Sources are available at https://github.com/schaefi/dice.
+
+.. code:: bash
+
+    $ sudo zypper in ruby2.1-rubygem-dice
+
+Installing and Setting up docker
```
Comment (tomschr) [inline doc/source/container_builder.rst]: docker -> Docker

File: doc/source/container_builder.rst [hunk 5]
```diff
@@ -0,0 +1,276 @@
...
+--------------------------------
+
+docker packages are usually available with the used distribution
+
+.. code:: bash
+
+    $ sudo zypper in docker
+
+Make sure that the user who is intended to build images is a member
+of the `docker` group by running the following command:
```
Comment (tomschr) [inline doc/source/container_builder.rst]: Rephrase it like:

```
Make sure that the user, who is intended to build images, is a member
of the `docker` group. Run the following command:
```

File: doc/source/container_builder.rst [hunk 6]
```diff
@@ -0,0 +1,276 @@
...
+    $ sudo useradd -G docker <builduser>
+
+It is required to logout and login again to let this change
+become active.
+
+Once this is done you need to setup the Docker storage backend.
+By default Docker uses the device mapper to manage the storage for
+the containers it starts. Unfortunately, this does not work if the
+container is supposed to build images because it runs into conflicts
+with tools like kpartx which itself maps devices using the device
```
Comment (tomschr) [inline doc/source/container_builder.rst]: Is kpartx a command line tool? Use `:command:`?

File: doc/source/container_builder.rst [hunk 7]
```diff
@@ -0,0 +1,276 @@
...
+mapper.
+
+Fortunately, there is a solution for Docker which allows us to use
+btrfs as the storage backend. The following is only required if your
+host system root filesystem is not btrfs.
```
Comment (tomschr) [inline doc/source/container_builder.rst]: btrfs -> Btrfs as by our styleguide ;-)

File: doc/source/container_builder.rst [hunk 8]
```diff
@@ -0,0 +1,276 @@
...
+
+      DOCKER_OPTS="-s btrfs"
+
+Finally start the docker daemon:
+
+.. code:: bash
+
+    $ sudo systemctl restart docker
+
+Installing and Setting up the build container
```
Comment (tomschr) [inline doc/source/container_builder.rst]: build container -> Build Container

File: doc/source/container_builder.rst [hunk 9]
```diff
@@ -0,0 +1,276 @@
...
+at https://build.opensuse.org/project/show/Virtualization:Appliances:Images
+On a regular basis the result image is pushed
+to https://hub.docker.com/r/schaefi/kiwi-build-box
+
+There are two ways to import the build container to your local docker system
+
+1. Download from the openSUSE Buildservice and manually import
+2. Use docker to pull the box from dockerhub
+
+Pull from dockerhub
```
Comment (tomschr) [inline doc/source/container_builder.rst]: dockerhub -> Dockerhub (or DockerHub?)

File: doc/source/container_builder.rst [hunk 10]
```diff
@@ -0,0 +1,276 @@
...
+-------------------
+
+.. code:: bash
+
+    $ docker pull schaefi/kiwi-build-box:latest
+
+Download from the open BuildService
```
Comment (tomschr) [inline doc/source/container_builder.rst]: open -> Open

File: doc/source/container_builder.rst [hunk 11]
```diff
@@ -0,0 +1,276 @@
...
+
+    This step can be skipped if there are no complex provision tasks
+    of the building environment required.
+
+Installing vagrant is well documented at
+https://docs.vagrantup.com/v2/installation/index.html
+
+Access to a machine started by vagrant is done through ssh exclusively.
+Because of that an initial key setup is required in the box vagrant should
+start. The kiwi provided build boxes includes the public key of the vagrant
```
Comment (tomschr) [inline doc/source/container_builder.rst]: kiwi -> KIWI

File: doc/source/container_builder.rst [hunk 12]
```diff
@@ -0,0 +1,276 @@
...
+key pair and thus allows access. It is important to understand that the
+private vagrant key is not a secure key because the private key is not
+protected. However this is not a problem because vagrant creates a new
```
Comment (tomschr) [inline doc/source/container_builder.rst]: However -> However,

File: doc/source/container_builder.rst [hunk 13]
```diff
@@ -0,0 +1,276 @@
...
+key pair for each machine it starts. In order to allow vagrant the initial
+access and the creation of a new key pair, it's required to provide access
+to the insecure vagrant private key. The following commands should not be
+executed as root, but rather as the user intending to build images.
```
Comment (tomschr) [inline doc/source/container_builder.rst]: "as the user intending to build images." -> "the intended user to build images."

File: doc/source/container_builder.rst [hunk 14]
```diff
@@ -0,0 +1,276 @@
...
+    Dice.configure do |config|
+      config.buildhost = :DOCKER
+    end
+
+Building with dice
+------------------
+
+If you have choosen to just use the default dice configuration as
+provided with the example appliance descriptions, the following example
+command will build the image.
```
Comment (tomschr) [inline doc/source/container_builder.rst]: Missing colon at the end of "image".

File: doc/source/container_builder.rst [hunk 15]
```diff
@@ -0,0 +1,276 @@
...
+
+.. code:: bash
+
+    $ cd <git-clone-result-kiwi-descriptions>
+
+    $ dice build suse/x86_64/suse-leap-42.1-JeOS
+    $ dice status suse/x86_64/suse-leap-42.1-JeOS/
+
+
+Buildsystem backends
```
Comment (tomschr) [inline doc/source/container_builder.rst]: backends -> Backends

File: doc/source/container_builder.rst [hunk 16]
```diff
@@ -0,0 +1,276 @@
...
+--------------------
+
+dice currently supports three build system backends:
+
+1. `host buildsystem` - dice builds on the host like if you would call
+   kiwi on the host directly
```
Comment (tomschr) [inline doc/source/container_builder.rst]: kiwi -> KIWI

File: doc/source/container_builder.rst [hunk 17]
```diff
@@ -0,0 +1,276 @@
...
+
+2. `vagrant buildsystem` - dice uses vagrant to run a virtual system which
```
Comment (tomschr) [inline doc/source/container_builder.rst]: Depending on context, do you mean Vagrant (as a project?) or `command:`vagrant` as a command?

File: doc/source/container_builder.rst [hunk 18]
```diff
@@ -0,0 +1,276 @@
...
+   could also be a container and build the image on this machine
+
+3. `docker buildsystem` - dice uses docker directly to run the build in
```
Comment (tomschr) [inline doc/source/container_builder.rst]: Same here, docker vs. `:command:docker`

File: doc/source/container_builder.rst [hunk 19]
```diff
@@ -0,0 +1,276 @@
...
+   a container
+
+So far we have described how to use dice with the plain docker
```
Comment (tomschr) [inline doc/source/container_builder.rst]: Missing comma: So far, ...

File: doc/source/container_builder.rst [hunk 20]
```diff
@@ -0,0 +1,276 @@
...
+buildsystem. If the build task requires additional content or logic
+before the build can start the vagrant buildsystem configured to use
+docker provides a nice interface to this provisioning tasks.
+
+Building with the vagrant buildsystem
```
Comment (tomschr) [inline doc/source/container_builder.rst]: Title style: Building with the Vagrant Buildsystem

File: doc/source/container_builder.rst [hunk 21]
```diff
@@ -0,0 +1,276 @@
...
+-------------------------------------
+
+The following sections describes how to setup dice to use docker in
+combination with vagrant as provisioning system
```
Comment (tomschr) [inline doc/source/container_builder.rst]: Some minor typos (depending on if you want to refer to the command or the project):

```
The following sections describes how to setup dice to use Docker in
combination with Vagrant as provisioning system.
```

File: doc/source/container_builder.rst [hunk 22]
```diff
@@ -0,0 +1,276 @@
...
+
+The Dicefile
+------------
+
+The Dicefile in the context of vagrant needs to know the username to
+access the container. This is because in vagrant access to the system
+is handled over ssh. vagrant is also the default buildsystem in dice
+which means in contrast to the docker buildsystem we do not have to
+actively select it.
```
Comment (tomschr) [inline doc/source/container_builder.rst]: Some suggestions (not sure if it catches the sense):

```
The Dicefile in the context of Vagrant needs to know the user name to
access the container. The reason for this is, in Vagrant access to the
system is handled over SSH. As Vagrant is also the default build system
in Dice, it is automatically selected [in ....]. In contract to the Docker build system,
Vagrant needs to be manually selected.
```

File: doc/source/container_builder.rst [hunk 23]
```diff
@@ -0,0 +1,276 @@
...
+.. code:: ruby
+
+    Dice.configure do |config|
+      config.ssh_user = "vagrant"
+    end
+
+The Vagrantfile
+---------------
+
+Once you call dice to build the image it will call vagrant to bring up the
```
Comment (tomschr) [inline doc/source/container_builder.rst]: Use `:command:` around "vagrant".

File: doc/source/container_builder.rst [hunk 24]
```diff
@@ -0,0 +1,276 @@
...
+container. In order to allow vagrant to do this we have to tell vagrant that
+it should use Docker for this task and provide parameters on how to run the
+container. At the same place the Dicefile exists we create the Vagrantfile
```
Comment (tomschr) [inline doc/source/container_builder.rst]: Suggestion:

```
In order to allow this, we have to tell Vagrant to use Docker for this task and provide
parameters on how to run the container.
```

File: doc/source/container_builder.rst [hunk 25]
```diff
@@ -0,0 +1,276 @@
...
+  and control virtual machines and container instances. vagrant
+  has a very nice way to provision a machine prior to running the
+  actual build. vagrant also supports docker as a provider which
+  makes it a perfect fit for complex provisioning tasks in
+  combination with the docker container system.
+
+Installing and Setting up dice
+------------------------------
+
+dice packages are available at the respective buildservice project
```
Comment (tomschr) [inline doc/source/container_builder.rst]: I would rephrase that to:

The dice packages and sources are available at the following locations:

* Build service project: http://download.opensuse.org/repositories/Virtualization:/Appliances
* Sources: https://github.com/schaefi/dice

File: doc/source/container_builder.rst [hunk 26]
```diff
@@ -0,0 +1,276 @@
+Building in a self-contained Environment
```
Comment (tomschr) [inline doc/source/container_builder.rst]: Title style: self-contained -> Self-Contained

**Comment by tomschr** [comment]
@schaefi Ok, I did a quick review, see my suggestions.

Apart from this, I have some consistency concerns. ;) It is not always clear if you refer to Docker as a project (would be capital "D") or `docker` as a command (executed by the user). The same applies to Dice/`dice` and Vagrant/`vagrant`.

Hope this helps. Probably another round would be helpful too. :))

File: doc/source/container_builder.rst (hunk 3 above)
Comment (schaefi) [inline doc/source/container_builder.rst]: it should be "run, provision and control virtual machines and container instances."

**Comment by schaefi** [comment]
Thanks much, I'll care for the scope of the names and will include all your suggestions asap

**Comment by schaefi** [comment]
All right first round of editorial changes are done

Thanks much

**Comment by tomschr** [comment]
@schaefi Ok, great! For the first round, I've focused more on grammar, typo, and consistency issues.

Should I review it again and focus on content?

**Comment by schaefi** [comment]
If you have the time that would be great

**Comment by tomschr** [comment]
@schaefi Ok, will do. :)

File: doc/source/container_builder.rst [hunk 27]
```diff
@@ -0,0 +1,282 @@
+Building in a Self-Contained Environment
+========================================
+
+.. topic:: Abstract
+
+   Users building images with KIWI face problems if they want
+   to build as non root user or can't install KIWI on their
+   system due to inappropriate build host requirements or
+   project policies.
```
Comment (tomschr) [inline doc/source/container_builder.rst]: I'm not sure if I understood the above sentence correctly. So there are two problems, right? This little "or" is a bit confusing. Maybe better start with this:

```Users building images with KIWI face two problems: 

```

File: doc/source/container_builder.rst [hunk 28]
```diff
@@ -0,0 +1,282 @@
...
+
+   This document describes how to perform the build process in
+   a Docker container using the Dice containment build system
+   written for KIWI.
+
+   The changes to the machine to become a build host will
```
Comment (tomschr) [inline doc/source/container_builder.rst]: to the machine -> on the machine(?) not sure about that.

File: doc/source/container_builder.rst [hunk 29]
```diff
@@ -0,0 +1,282 @@
...
+------------
+
+The following components needs to be installed on the build system:
+
+* Dice - a containment build system for KIWI.
+
+* Docker - a container framework based on the Linux
+  container support in the kernel.
+
+* Docker Build Container for KIWI.
```
Comment (tomschr) [inline doc/source/container_builder.rst]: Isn't it a Docker image? Shouldn't it be something like:

```
* Docker Image - a Docker build container for KIWI
```

It would be consistent with the other items in the list (first the term, dash, and then the explanation).

File: doc/source/container_builder.rst [hunk 30]
```diff
@@ -0,0 +1,282 @@
...
+    $ sudo zypper in ruby2.1-rubygem-dice
+
+Installing and Setting up Docker
+--------------------------------
+
+Docker packages are usually available with the used distribution.
+
+.. code:: bash
+
+    $ sudo zypper in docker
```
Comment (tomschr) [inline doc/source/container_builder.rst]: Should we distinguish here between the different distributions? Probably they are named the same (`docker`?) on all systems?

File: doc/source/container_builder.rst [hunk 31]
```diff
@@ -0,0 +1,282 @@
...
+.. code:: bash
+
+    $ sudo systemctl restart docker
+
+Installing and Setting up the Build Container
+----------------------------------------------
+
+In order to build in a contained environment Docker has to start a
+privileged system container. Such a container must be imported before
+Docker can use it. The Build Container is provided to you as a
```
Comment (tomschr) [inline doc/source/container_builder.rst]: I guess, "Build Container" should be "build container".

File: doc/source/container_builder.rst [hunk 32]
```diff
@@ -0,0 +1,282 @@
...
+service and build with KIWI in the project
+at https://build.opensuse.org/project/show/Virtualization:Appliances:Images
+On a regular basis the result image is pushed
```
Comment (tomschr) [inline doc/source/container_builder.rst]: What does "On a regular basis" actually means? Is there also an irregular basis? ;-)

Maybe simplify by just writing: "The result image is pushed to ..."

File: doc/source/container_builder.rst [hunk 33]
```diff
@@ -0,0 +1,282 @@
...
+to https://hub.docker.com/r/opensuse/dice.
+
+There are two ways to import the Build Container to your local Docker system
+
+1. Download from the openSUSE Buildservice and manually import
+2. Use :command:`docker` to pull the box from Dockerhub
```
Comment (tomschr) [inline doc/source/container_builder.rst]: to "pull the box"? What does "box" means here? You mean probably "pull the image" from DockerHub, right?

However, I see a potential conflict of terms between the term "image": a "Docker image" and a "KIWI image" is probably not the same. In that case, better always state which image you mean.

If you agree, better write "pull the Docker image".

File: doc/source/container_builder.rst [hunk 34]
```diff
@@ -0,0 +1,282 @@
...
+Download from the Open BuildService
+-----------------------------------
+
+Download the .tar.bz2 file which starts with :file:`Docker-Tumbleweed`
+
+.. code:: bash
+
+    $ wget http://download.opensuse.org/repositories/Virtualization:/Appliances:/Images/images/Docker-Tumbleweed.XXXXXXX.docker.tar.xz
+
+Import the downloaded tarball to docker as follows:
```
Comment (tomschr) [inline doc/source/container_builder.rst]: to docker -> with :command:`docker` as follows:

File: doc/source/container_builder.rst [hunk 35]
```diff
@@ -0,0 +1,282 @@
...
+    $ cat Docker-Tumbleweed.XXXXXXX.docker.tar.xz | docker import - opensuse/dice:latest
+
+
+Installing and Setting up Vagrant
+---------------------------------
+
+.. note:: Optional step
+
+    This step can be skipped if there are no complex provision tasks
+    of the building environment required.
```
Comment (tomschr) [inline doc/source/container_builder.rst]: Depends on what is a "complex provision task". What qualifies a task to be "complex"? :))

File: doc/source/container_builder.rst [hunk 36]
```diff
@@ -0,0 +1,282 @@
...
+
+Installing Vagrant is well documented at
+https://docs.vagrantup.com/v2/installation/index.html
+
+Access to a machine started by Vagrant is done through ssh exclusively.
```
Comment (tomschr) [inline doc/source/container_builder.rst]: ssh -> SSH

File: doc/source/container_builder.rst [hunk 37]
```diff
@@ -0,0 +1,282 @@
...
+   project policies.
+
+   This document describes how to perform the build process in
+   a Docker container using the Dice containment build system
+   written for KIWI.
+
+   The changes to the machine to become a build host will
+   be reduced to the requirements of Dice and Docker.
+
+Requirements
```
Comment (tomschr) [inline doc/source/container_builder.rst]: Although I have a some understanding about KIWI, I stumbled over some terms which I never heard of ("box/boxes"). Maybe a short terminology section could make sense here or at the end of this article?

File: doc/source/container_builder.rst [hunk 38]
```diff
@@ -0,0 +1,282 @@
...
+
+Installing Vagrant is well documented at
+https://docs.vagrantup.com/v2/installation/index.html
+
+Access to a machine started by Vagrant is done through ssh exclusively.
+Because of that an initial key setup is required in the box vagrant should
+start. The KIWI build boxes includes the public key of the Vagrant
+key pair and thus allows access. It is important to understand that the
+private Vagrant key is not a secure key because the private key is not
+protected. However, this is not a problem because Vagrant creates a new
```
Comment (tomschr) [inline doc/source/container_builder.rst]: I would introduce a new paragraph between "protected." and "However".

File: doc/source/container_builder.rst [hunk 39]
```diff
@@ -0,0 +1,282 @@
...
+
+The descriptions hosted here also provides a default :file:`Dicefile`
+as part of each image description.
+
+The Dicefile
+------------
+
+The Dicefile is the configuration file for the dice buildsystem backend.
+All it needs to know for a plain docker based build process is the
+selection of the buildhost to be a docker container. The Dicefile's
```
Comment (tomschr) [inline doc/source/container_builder.rst]: docker container -> Docker container

File: doc/source/container_builder.rst [hunk 40]
```diff
@@ -0,0 +1,282 @@
...
+
+    Dice.configure do |config|
+      config.ssh_user = "vagrant"
+    end
+
+The Vagrantfile
+---------------
+
+The existence of a Vagrantfile tells Dice to use Vagrant as Buildsystem.
+Once you call dice to build the image it will call :command:`vagrant` to
```
Comment (tomschr) [inline doc/source/container_builder.rst]: Wrap `:command:` around the word `dice`.

**Comment by tomschr** [comment]
@schaefi Ok, I think I'm through. To be honest, I'm not sure if I understood everything. :) I'm distracted with other things at the moment, so probably it's just me. :)

If you have any further questions, let me know.

File: doc/source/container_builder.rst (hunk 27 above)
Comment (schaefi) [inline doc/source/container_builder.rst]: Well the 'or' is correct, there are users who have issues to run a program as root, others have issues installing KIWI on their system because it does not exist for this distribution, and then again there are users who have issues with both. I think saying that everybody is facing all of these issues wouldn't be correct ?

If only one of the 'or' list applies it makes sense to read this article, that's what should be expressed

File: doc/source/container_builder.rst (hunk 30 above)
Comment (schaefi) [inline doc/source/container_builder.rst]: yes it's just docker everywhere

File: doc/source/container_builder.rst (hunk 35 above)
Comment (schaefi) [inline doc/source/container_builder.rst]: That really depends on the customer. So far dice only shares the image description directory with the docker image, if you need more than that to build, the vagrant provisioning makes sense. Maybe it should be said that way ?

File: doc/source/container_builder.rst (hunk 37 above)
Comment (schaefi) [inline doc/source/container_builder.rst]: box = "docker image" I will change that consistently

**Comment by schaefi** [comment]
Thanks for the review. It would be really good if the article is clear enough so that the message

"building an image in a container makes the build process independent of the host system"

gets caught in the mind of the readers, then when they stumble over the problem, Oh I need to be
root, or Oh kiwi does not exist on my Debian System, they should remember this article and give
Dice a try

If this is not clear to you, I think the article needs to be changed at the toplevel
